In [ ]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt 
import seaborn as sns 
import sklearn
import regex as re
import nltk
nltk.download('stopwords')
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.tag import pos_tag
nltk.download('averaged_perceptron_tagger')
from nltk.tokenize import RegexpTokenizer
from nltk.tag import PerceptronTagger
from collections import Counter
from nltk.stem import PorterStemmer

!python -m nltk.downloader punkt
from google.colab import drive
drive.mount('/content/drive')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


/usr/lib/python3.8/runpy.py:127: RuntimeWarning: 'nltk.downloader' found in sys.modules after import of package 'nltk', but prior to execution of 'nltk.downloader'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
data = pd.read_csv('../content/drive/MyDrive/Colab Notebooks/h_artists_songs.csv') 

In [ ]:
lyrics = data['Lyrics'].tolist()

In [ ]:
lyrics = [re.sub('\s+', ' ', text) for text in lyrics]

# removing non-English letters from lyrics
lyrics = [re.sub('[^a-zA-Z0-9]+', ' ',text) for text in lyrics]

# removing word 'Romanized' from lyrics' text
lyrics = [re.sub('Romanized', ' ', text) for text in lyrics]



#Part1

In [ ]:
tokenized_lyrics = []
tokenizer = RegexpTokenizer(r'\w+')
for song in lyrics:
  tokenized_lyrics.extend(tokenizer.tokenize(song.lower()))

counter = Counter(tokenized_lyrics)

top_20_tokens = counter.most_common(20)

# Print the top 20 most frequent tokens and their frequencies
print(top_20_tokens)

[('i', 398111), ('you', 305416), ('the', 285885), ('to', 169739), ('and', 166153), ('a', 153805), ('it', 140690), ('me', 136123), ('t', 128903), ('s', 121650), ('my', 108543), ('in', 103929), ('m', 87304), ('that', 84415), ('of', 78766), ('we', 76239), ('on', 75463), ('your', 72928), ('all', 63774), ('is', 56276)]


In [ ]:
tokenized_lyrics_no_stop = []
stop_words = stopwords.words('english')
for token in tokenized_lyrics:
  if token not in stop_words:
    tokenized_lyrics_no_stop.append(token)
porter = PorterStemmer()
stems = [porter.stem(word) for word in tokenized_lyrics_no_stop]
top_20_stems = Counter(stems).most_common(20)
print(top_20_stems)

[('love', 57557), ('know', 51731), ('like', 44941), ('oh', 41254), ('go', 34454), ('get', 32530), ('got', 30926), ('time', 29214), ('one', 28595), ('come', 28511), ('let', 27234), ('yeah', 26021), ('never', 25781), ('feel', 24545), ('want', 24459), ('see', 24286), ('make', 23450), ('la', 23307), ('way', 22793), ('say', 22412)]


In [ ]:
tagger = nltk.PerceptronTagger()
tags = tagger.tag(tokenized_lyrics_no_stop)

verbs = [word.lower() for word, tag in tags if tag.startswith('V')]
verb_counts = Counter(verbs)

top_verbs = verb_counts.most_common(20)
print(top_verbs)

[('know', 39155), ('got', 29575), ('go', 27172), ('love', 22096), ('get', 21752), ('say', 18995), ('see', 17488), ('make', 16546), ('want', 16519), ('take', 16112), ('come', 15842), ('let', 13540), ('said', 9966), ('think', 9028), ('keep', 8800), ('need', 8724), ('feel', 7886), ('gone', 7385), ('tell', 7280), ('find', 7069)]


##Were you surprised by the results? Why or why not?##
For the top 20 most frequent tokens, we were surprised that we got a lot of single letters such as “I”, “a”. We were hesitant whether to include the stop words. After discussing the interpretations of the results we got, we decided to keep the stop words. With the stop words, we can learn better what the most common tokens are in songs. And those tokens could also be the most common used in English sentences. The verb list later show more meaningful items such as know, go, love, say. These verbs may represent the most frequent used verbs by singers. 

# Part2

In [ ]:
data2022=data[data['Year']==2022].reset_index(drop=True)

In [ ]:
lyrics2022 = data2022['Lyrics'].tolist()

lyrics2022 = [re.sub('\s+', ' ', text) for text in lyrics2022]

lyrics2022 = [re.sub('[^a-zA-Z0-9]+', ' ',text) for text in lyrics2022]

lyrics2022 = [re.sub('Romanized', ' ', text) for text in lyrics2022]


In [ ]:
tokenized_lyrics22 = []
tokenizer = RegexpTokenizer(r'\w+')
for song in lyrics2022:
  tokenized_lyrics22.extend(tokenizer.tokenize(song.lower()))

counter22 = Counter(tokenized_lyrics22)

top_20_tokens22 = counter22.most_common(20)

# Print the top 20 most frequent tokens and their frequencies
print(top_20_tokens22)

[('i', 10305), ('you', 7270), ('the', 6473), ('to', 3746), ('it', 3563), ('and', 3241), ('a', 3231), ('t', 3193), ('me', 3046), ('my', 2732), ('s', 2634), ('in', 2501), ('m', 2331), ('we', 2088), ('that', 1955), ('of', 1777), ('all', 1743), ('your', 1670), ('on', 1591), ('don', 1468)]


In [ ]:
tokenized_lyrics_no_stop22 = []
stop_words = stopwords.words('english')
for token in tokenized_lyrics22:
  if token not in stop_words:
    tokenized_lyrics_no_stop22.append(token)
porter22 = PorterStemmer()
stems22 = [porter22.stem(word) for word in tokenized_lyrics_no_stop22]
top_20_stems22 = Counter(stems22).most_common(20)
print(top_20_stems22)

[('know', 1430), ('like', 1250), ('oh', 1244), ('love', 1106), ('go', 811), ('yeah', 783), ('get', 783), ('never', 741), ('one', 721), ('got', 709), ('let', 695), ('want', 654), ('time', 637), ('come', 631), ('way', 619), ('feel', 605), ('back', 596), ('take', 586), ('see', 538), ('wanna', 517)]


In [ ]:
tagger22 = nltk.PerceptronTagger()
tags22 = tagger22.tag(tokenized_lyrics_no_stop22)

verbs22 = [word.lower() for word, tag in tags22 if tag.startswith('V')]
verb_counts22 = Counter(verbs22)

top_verbs22 = verb_counts22.most_common(20)
print(top_verbs22)

[('know', 1123), ('got', 673), ('go', 615), ('get', 457), ('want', 445), ('say', 428), ('take', 422), ('love', 411), ('let', 404), ('see', 382), ('make', 358), ('come', 311), ('keep', 281), ('think', 262), ('feel', 233), ('find', 225), ('tell', 215), ('said', 209), ('need', 209), ('run', 173)]


##In a paragraph, compare your results from Part 2 to those from Part 1. How much overlap is there?##

Even if we filtered by genre, the most common tokens are still the same, which made us think that the most common tokens are prevalent in all English songs regardless of genre or date. For the list of stems and verbs, love appears less. As in the genre of 2022, singers may not focus so much on romantic. It can be due to the war or the pandemic. 

#Part3

In [ ]:
Hagg=data[data["Artist_Name"]=='Haggard, Merle'].reset_index(drop=True)
Jr=data[data["Artist_Name"]=='Hank Williams Jr.'].reset_index(drop=True)

In [ ]:
lyrics_hagg = Hagg['Lyrics'].tolist()
lyrics_hagg = [re.sub('\s+', ' ', text) for text in lyrics_hagg]
lyrics_hagg = [re.sub('[^a-zA-Z0-9]+', ' ',text) for text in lyrics_hagg]
lyrics_hagg = [re.sub('Romanized', ' ', text) for text in lyrics_hagg]

tokenized_lyrics_hagg = []
tokenizer = RegexpTokenizer(r'\w+')
for song in lyrics_hagg:
  tokenized_lyrics_hagg.extend(tokenizer.tokenize(song.lower()))
counter_hagg = Counter(tokenized_lyrics_hagg)

tokenized_lyrics_no_stop_hagg = []
stop_words = stopwords.words('english')
for token in tokenized_lyrics_hagg:
  if token not in stop_words:
    tokenized_lyrics_no_stop_hagg.append(token)
porter_hagg = PorterStemmer()
stems_hagg = [porter_hagg.stem(word) for word in tokenized_lyrics_no_stop_hagg]
top_20_stems_hagg = Counter(stems_hagg).most_common(20)
print(top_20_stems_hagg)

stems_hagg=' '.join(stems_hagg)

[('love', 892), ('time', 550), ('know', 520), ('old', 422), ('go', 394), ('got', 392), ('like', 383), ('come', 378), ('one', 372), ('way', 357), ('man', 342), ('home', 326), ('day', 307), ('never', 283), ('make', 283), ('back', 276), ('let', 271), ('blue', 266), ('good', 262), ('alway', 254)]


In [ ]:
lyrics_jr = Jr['Lyrics'].tolist()
lyrics_jr = [re.sub('\s+', ' ', text) for text in lyrics_jr]
lyrics_jr = [re.sub('[^a-zA-Z0-9]+', ' ',text) for text in lyrics_jr]
lyrics_jr = [re.sub('Romanized', ' ', text) for text in lyrics_jr]

tokenized_lyrics_jr = []
tokenizer = RegexpTokenizer(r'\w+')
for song in lyrics_jr:
  tokenized_lyrics_jr.extend(tokenizer.tokenize(song.lower()))
counter_jr = Counter(tokenized_lyrics_jr)

tokenized_lyrics_no_stop_jr = []
stop_words = stopwords.words('english')
for token in tokenized_lyrics_jr:
  if token not in stop_words:
    tokenized_lyrics_no_stop_jr.append(token)
porter_jr = PorterStemmer()
stems_jr = [porter_jr.stem(word) for word in tokenized_lyrics_no_stop_jr]
top_20_stems_jr = Counter(stems_jr).most_common(20)
print(top_20_stems_jr)
stems_jr=' '.join(stems_jr)

[('love', 767), ('know', 665), ('got', 650), ('like', 610), ('get', 506), ('go', 470), ('one', 405), ('caus', 404), ('time', 402), ('man', 366), ('gonna', 360), ('never', 360), ('well', 345), ('come', 324), ('make', 306), ('take', 304), ('heart', 298), ('old', 296), ('oh', 289), ('way', 289)]


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

tfidf_vectorizer = TfidfVectorizer(stop_words='english')

tfidf_matrix1 = tfidf_vectorizer.fit_transform([stems_jr])
tfidf_matrix2 = tfidf_vectorizer.transform([stems_hagg])

cosine_similarities = cosine_similarity(tfidf_matrix1, tfidf_matrix2)

cosine_similarities

array([[0.93176232]])

##Discuss why the two songs were most alike. What is this analysis able to tell us?

We got 0.93176, which suggests their songs are very similar based on the stems. Merle Haggard and Hank Williams Jr. are known for their storytelling abilities and often used their music to explore a range of topics related to life, love, and the human experience. From the top 20 stems of their songs, we can also see the "love" and "know" are very common. 